src/main/java/com/fasterxml/jackson/databind/jsontype/impl/StdTypeResolverBuilder.java

```diff
@@ -129,6 +129,8 @@ public TypeDeserializer buildTypeDeserializer(DeserializationConfig config,
             //   seems like a reasonable compromise.
             if (_defaultImpl == Void.class) {
                 defaultImpl = config.getTypeFactory().constructType(_defaultImpl);
+            } else if (!baseType.getRawClass().isAssignableFrom(_defaultImpl)) {
+                defaultImpl = null;
```

Comment:
If the default implementation is not a subtype of base type, should this not rather throw an exception?
Otherwise it will simply hide incorrect usage.
(I realize that ideally method should allow throwing of `JsonMappingException`, or just `IOException`, so for now that'd have to be something like `IllegalArgumentException` that will get converted up the call stack).

Reply:
In this baseType is a target type of deserialization, AFAIK.
Let's say we have abstract class Animal, implementations are Dog, Cat etc and UnknownAnimal as default implementation.
In some file (e.g. cats.json) I have json with Cat and no other animals. In this case it is suitable to parse it as Cat. In current implementation code like this will throw an exception because UnknownAnimal is not subclass of Cat (it's subclass of Animal):
Cat cat = new ObjectMapper().readValue(json, Cat.class);
Ugly workaround is to use class cast:
Cat cat = (Cat) new ObjectMapper().readValue(json, Animal.class);
And it's even worse if I need to parse a list of cats:
List<Cat> cat = (List<Cat>) new ObjectMapper().readValue(json, new TypeReference<List<Animal>>() {});
Instead of:
List<Cat> cat = new ObjectMapper().readValue(json, new TypeReference<List<Cat>>() {});
As for me parsing specific subtype is absolutely correct behavior if input data is correct and has only objects of specific type. And yes, it should throw an exception if input json contains other type like Dog or some UnknownAnimal. But it may not throw an exception if all input data is correct and can be parsed as or cast to target type. 

BTW, this functionality worked in version 2.3.5, but was broken in some later versions, I've checked 2.7 and 2.9.

Reply:
@masyamandev Sounds good. I am hoping to work a bit on this too -- I think your patch helped a lot, showing that fix could be quite simple, fundamentally. I hadn't had time to dig deeper into this, although I am aware that this is a relatively common problem.

Reply:
I have added a couple of tests which checks DeserializerCache by deserializing different variations of objects and types using single ObjectMapper. It seems that it works fine.

Regarding throwing an Exception when default implementation class does not extends base class (UnknownAnimal is not a subclass of Animal). Do you think it's a critical issue? 
There are too many cases when Exception can be thrown in runtime when parsing objects. For instance:
- new ObjectMapper().readValue("{\"type\":\"dog\"}", Cat.class);
- new ObjectMapper().readValue("{\"type\":\"unknownAnimal\"}", Cat.class);
- new ObjectMapper().readValue("{\"type\":\"cat\"}", SphinxCat.class); // SphinxCat extends Cat
- new ObjectMapper().readValue("{\"type\":\"unknownAnimal\"}", Animal.class); // In case if UnknownAnimal does not extend Animal
In proposed implementation Exception will not be thrown for cases when UnknownAnimal does not extend Animal:
- new ObjectMapper().readValue("{\"type\":\"cat\"}", Animal.class)
- new ObjectMapper().readValue("{\"type\":\"cat\"}", Cat.class)
But will be thrown when we try to parse "{\"type\":\"unknownAnimal\"}.
Unfortunately I can't find simple solution where to add checking root type and default implementation compatibility.

Reply:
BTW, I've found another PR which fixes exactly the same issue.
https://github.com/FasterXML/jackson-databind/pull/1656
So I'm not the only one who needs this.
That PR contains code for checking root type. But it has slightly smaller amount of tests. 
I can implement checking root type the same way as in PR 1656 into my PR if it's acceptable solution.

Reply:
@masyamandev I think what would be great would be to start with unit tests: ones that already pass can be added wherever they make sense; and ones failing now (but to be fixed) under `failing`. I think we agree on definition on what should work, and perhaps an what shouldn't too.

But on latter, just to make sure, I'd want to check the case where annotation defined incompatible type. Something like:

```java
public class Zoo {
   @JsonTypeInfo(.... defaultImpl = NotAnimal.class)
   public Animal animal;
}
```

I am not as worried about what root type is being requested, or whether incoming type id is compatible, just because those cases are already covered (... as far as I know).
I really just want to prevent obviously incompatible case.

Now: ideally this would be caught by annotation processing logic, but unfortunately this is not currently (2.x) possible, as annotations are first flattened throughout type hierarchy, and it is not possible to know what class actually had annotations.
If this was available, we could probably simply skip the checks.